### PR TITLE
[CIT-697] Allowing colon in datetime deserialisation

### PR DESCRIPTION
**Pull request by initard**
Signed-off-by: initard <solo@softwareag.com>

## Proposed changes

C8y smartrest payload returns datetime as follows:

```shell
c8y result: 2021-10-23T19:03:26+0100
```
However, chrono expects the following datetime:
```shell
rfc3339 expected: 2021-10-23T19:03:26+01:00
```
The only difference being the colon (:) at the end. This fix allows both types of date formats to be parsed.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://cumulocity.atlassian.net/browse/CIT-612

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs [hunk 1]
```diff
@@ -178,9 +178,12 @@ where
     // so we add a ':'
     let mut date_string: String = Deserialize::deserialize(deserializer)?;
     let str_size = date_string.len();
-    date_string = date_string[0..str_size - 2].to_string()
-        + ":"
-        + &date_string[str_size - 2..str_size].to_string();
+    // check if `date_string` does not have a colon.
+    if !date_string.contains(':') {
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs:182]: You should only look for the 3rd occurrence, others are not an issue.

Comment (initard) [reply mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs]: Yes thank you!

**Comment by initard** [comment]
Marked as draft as it seems the "fix" didn't work as expected.

File: mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs [hunk 2]
```diff
@@ -178,9 +178,17 @@ where
     // so we add a ':'
     let mut date_string: String = Deserialize::deserialize(deserializer)?;
     let str_size = date_string.len();
-    date_string = date_string[0..str_size - 2].to_string()
-        + ":"
-        + &date_string[str_size - 2..str_size].to_string();
+    // check if `date_string` does not have a colon.
```
Comment (didier-wenzek) [inline mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs:181]: The correct statement is *check that the time zone is formatted using a colon*.

**Review by didier-wenzek — APPROVED** [review]
Approved

File: mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs [hunk 3]
```diff
@@ -178,9 +178,17 @@ where
...
+    let date_string_end = &date_string.split('+').last();
+    date_string = match date_string_end {
+        Some(string) if !string.contains(':') => {
+            date_string[0..str_size - 2].to_string()
+                + ":"
+                + &date_string[str_size - 2..str_size].to_string()
+        }
+        _ => date_string,
+    };
+
```
Comment (PradeepKiruvale) [inline mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs:191]: Isn't it a good idea to add a unit test to check this?

Comment (didier-wenzek) [reply mapper/cumulocity/c8y_smartrest/src/smartrest_deserializer.rs:191]: Yes it is.

**Pull request merged**